### PR TITLE
Use the node-level datatypes API

### PR DESCRIPTION
**Pull request by ajreynol**
This updates the theory solvers to use the node-level datatypes API.

It also deletes a few obsolete comments about features that do not currently plan to be implemented: subtyping for tuples, and extensions to the skolemize module for more complex induction schemas.

### CODE REVIEW

File: src/theory/datatypes/datatypes_rewriter.cpp [hunk 1]
```diff
@@ -445,7 +441,7 @@ RewriteResponse DatatypesRewriter::rewriteTester(TNode in)
     return RewriteResponse(REWRITE_DONE,
                            NodeManager::currentNM()->mkConst(result));
   }
-  const Datatype& dt = static_cast<DatatypeType>(in[0].getType().toType()).getDatatype();
+  const DType& dt = in[0].getType().getDType();
```
Comment (HanielB) [inline src/theory/datatypes/datatypes_rewriter.cpp]: Just seeing this makes me so happy

File: src/theory/datatypes/theory_datatypes_type_rules.h [hunk 2]
```diff
@@ -104,22 +105,6 @@ struct DatatypeConstructorTypeRule {
         return false;
       }
     }
-    //if we support subtyping for tuples, enable this
```
Comment (HanielB) [inline src/theory/datatypes/theory_datatypes_type_rules.h]: Say in the PR description why you can delete this code?

File: src/theory/quantifiers/skolemize.cpp [hunk 3]
```diff
@@ -104,32 +104,17 @@ void Skolemize::getSelfSel(const Datatype& dt,
...
     {
-      TypeNode tn = TypeNode::fromType(dc[j].getRangeType());
+      TypeNode tn = dc[j].getRangeType();
       Trace("sk-ind-debug") << "Compare " << tn << " " << ntn << std::endl;
       if (tn == ntn)
       {
         ssc.push_back(n);
       }
     }
-    /* TODO: more than weak structural induction
```
Comment (HanielB) [inline src/theory/quantifiers/skolemize.cpp]: ditto

File: src/theory/datatypes/datatypes_rewriter.cpp (hunk 1 above)
Comment (ajreynol) [reply src/theory/datatypes/datatypes_rewriter.cpp]: Me too :)

File: src/theory/datatypes/theory_datatypes_type_rules.h (hunk 2 above)
Comment (ajreynol) [reply src/theory/datatypes/theory_datatypes_type_rules.h]: Done

**Comment by ajreynol** [comment]
Regarding performance tests, this actually still isn't the final product.  :)
The issue is that we are still running things through Datatype under the hood, see e.g. : https://github.com/CVC4/CVC4/blob/master/src/expr/node_manager.cpp#L265
Nevertheless I do think that we are gaining a lot of performance from this PR alone, will do a test and post soon.

**Comment by ajreynol** [comment]
```
                           Configuration  #unsat      time    #sat      time #solved    #total
          CVC4-120219-sac_sygus_d_status    2214   91406.9       0         0      2214      2855
           CVC4-120319-os_sygus_d_status    2215   92980.4       0         0      2215      2855
              CVC4-121219_sygus_d_status    2213    104972       0         0      2213      2855
          CVC4-121219-dti_sygus_d_status    2213    101082       0         0      2213      2855





For unsatisfiable : 
          CVC4-120219-sac_sygus_d_status       X      11      14      14
           CVC4-120319-os_sygus_d_status      12       X       6       9
              CVC4-121219_sygus_d_status      13       4       X       4
          CVC4-121219-dti_sygus_d_status      13       7       4       X

For satisfiable : 
          CVC4-120219-sac_sygus_d_status       X       0       0       0
           CVC4-120319-os_sygus_d_status       0       X       0       0
              CVC4-121219_sygus_d_status       0       0       X       0
          CVC4-121219-dti_sygus_d_status       0       0       0       X

```
Above `CVC4-121219-dti` is this PR vs master `CVC4-121219`, seems to be 3% faster. Notice that both are slower than an older version from 12/3/19.  Hopefully this will improve when everything is finalized.

**Pull request merged**